app/src/app/views/stix/collection/collection-import/collection-import-workflow/collection-import.component.ts

```diff
@@ -19,266 +22,544 @@ import { Tactic } from 'src/app/classes/stix/tactic';
-                                for (let category in results.import_categories) {
-                                    if (category == "errors") continue;
-                                    for (let id of results.import_categories[category]) if (!save_error_ids.has(id)) this.successfully_saved.add(id);
-                                }
-                                logger.log("Successfully imported the following objects:", Array.from(this.successfully_saved));
-                                this.stepper.next(); 
-                            },
-                            complete: () => { subscription.unsubscribe(); } //prevent memory leaks
-                        })
-                    })
-                }
-            },
-            complete: () => { promptSubscription.unsubscribe() } //prevent memory leaks
-        })
+    for (let object of collectionBundle.objects) {
+      // look up the category for the object
+      if (!(object.id in idToCategory)) {
+        // does not belong to a change category
+        this.unchanged_ids.push(object.id);
+        continue;
+      }
+      // track that this object has changed
+      this.changed_ids.push(object.id);
+      // determine the change category
+      let category = idToCategory[object.id];
+      // wrap the object as if it came from the back-end
+      let raw: { [key: string]: any } = { stix: object, workspace: {} };
+      // parse the object & add it to the appropriate category for rendering
+      switch (object.type) {
+        case 'attack-pattern': //technique
+          this.object_import_categories.technique[category].push(
+            new Technique(raw)
+          );
+          break;
+        case 'x-mitre-tactic': //tactic
+          this.object_import_categories.tactic[category].push(new Tactic(raw));
+          break;
+        case 'malware': //software
+        case 'tool':
+          this.object_import_categories.software[category].push(
+            new Software(object.type, raw)
+          );
+          break;
+        case 'relationship': //relationship
+          // build source and target objects if the source/target objects aren't being uploaded simultaneously (in case where only relationships are uploaded)
+          if (object.source_ref in idToSdo) {
+            raw.source_object = idToSdo[object.source_ref];
+          } else {
+            raw.source_object = {
+              stix: {
+                attackID: object.source_id,
+                created: object.created,
+                description: object.description,
+                id: object.source_ref,
+                modified: object.modified,
+                name: object.source_name,
+                spec_version: object.spec_version,
+                type: object.relationship_type,
+                external_references: [
+                  {
+                    source_name: 'mitre-attack',
+                    external_id: object.source_id,
+                  },
+                ],
```

Comment:
Similar comment here. In cases where the source object does not have an ATT&CK ID, the ATT&CK ID does not need an entry in the external references array.